### PR TITLE
feat(security_groups): add support for Security Group Rules

### PR DESCRIPTION
**Pull request by andrewjtait**

### CODE REVIEW

**Review by strideynet — APPROVED** [review]

File: core/security_group_rule.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+type SecurityGroupRuleRef struct {
+	ID string `json:"id,omitempty"`
+}
+
+func (sgr SecurityGroupRuleRef) queryValues() *url.Values {
+	v := &url.Values{}
+	v.Set("security_group_rule[id]", sgr.ID)
+
+	return v
```
Comment (jimeh) [inline core/security_group_rule.go:32]: Minor nitpick to simplify this like the `(SecurityGroupRef) queryValues()` method.

```suggestion
	return &url.Values{"security_group_rule[id]": []string{sgr.ID}}
```

File: core/security_group_rule.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+type SecurityGroupRuleArguments struct {
+	Direction string   `json:"direction,omitempty"`
+	Protocol  string   `json:"protocol,omitempty"`
+	Ports     string   `json:"ports,omitempty"`
+	Targets   []string `json:"targets,omitempty"`
+	Notes     string   `json:"notes,omitempty"`
```
Comment (jimeh) [inline core/security_group_rule.go:40]: Is it possible to specifically set `Targets` or `notes` to a empty list (remove existing targets), and empty string (clear out existing notes value)? If so these will need to be pointer types so we can differentiate between unchanged and clear updates.

File: core/security_group_rule.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func (s *SecurityGroupRulesClient) GetByID(
+	ctx context.Context,
+	id string,
+) (*SecurityGroupRule, *katapult.Response, error) {
+	return s.Get(ctx, SecurityGroupRuleRef{ID: id})
+}
+
+type SecurityGroupRuleCreateRequest struct {
```
Comment (jimeh) [inline core/security_group_rule.go:108]: This struct doesn't need to be exported, as it's only used for internal request purposes.

File: core/security_group_rule.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+	body, resp, err := s.doRequest(ctx, "POST", u, reqBody)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return body.SecurityGroupRule, resp, nil
+}
+
+type SecurityGroupRuleUpdateRequest struct {
```
Comment (jimeh) [inline core/security_group_rule.go:130]: This struct doesn't need to be exported, as it's only used for internal request purposes.

File: core/security_group_rule.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+
+type SecurityGroupRuleArguments struct {
+	Direction string   `json:"direction,omitempty"`
+	Protocol  string   `json:"protocol,omitempty"`
+	Ports     string   `json:"ports,omitempty"`
+	Targets   []string `json:"targets,omitempty"`
+	Notes     string   `json:"notes,omitempty"`
+}
+
+type SecurityGroupRulesResponseBody struct {
```
Comment (jimeh) [inline core/security_group_rule.go:43]: This struct doesn't need to be exported, as it's only used for internal request purposes.

File: core/security_group_rule_test.go [hunk 6]
```diff
@@ -0,0 +1,802 @@
...
+			} else {
+				assert.EqualError(t, err, tt.wantErr)
+			}
+
+			assert.Equal(t, tt.want, got)
+		})
+	}
+}
+
+func TestSecurityGroupRulesClient_List(t *testing.T) {
```
Comment (jimeh) [inline core/security_group_rule_test.go:384]: Very minor nitpick, but I like ordering the tests for different functions in the same order as the function themselves in the implementation file :)

File: core/security_group_rule.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+	Notes     string   `json:"notes,omitempty"`
+}
+
+func (sgr *SecurityGroupRule) Ref() SecurityGroupRuleRef {
+	return SecurityGroupRuleRef{ID: sgr.ID}
+}
+
+type SecurityGroupRuleRef struct {
+	ID string `json:"id,omitempty"`
+}
```
Comment (jimeh) [inline core/security_group_rule.go:26]: There's no JSON marshaling tests for this struct, though I've noticed that's actually an oversight missing from nearly every single `*Ref` type we have, and needs to be addressed. But let's add it for security group rules here to begin with :)

File: core/client.go [hunk 8]
```diff
@@ -54,6 +55,8 @@ func New(rm RequestMaker) *Client {
 		NetworkSpeedProfiles:            NewNetworkSpeedProfilesClient(rm),
 		Networks:                        NewNetworksClient(rm),
 		Organizations:                   NewOrganizationsClient(rm),
+		SecurityGroups:                  NewSecurityGroupsClient(rm),
```
Comment (jimeh) [inline core/client.go:58]: Setting up a new `SecurityGroupsClient` here probably belongs in #103 :)

**Review by jimeh — CHANGES_REQUESTED** [review]
I only have a few small things and some nitpicks, overall looks very good :)

File: core/security_group_rule.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+	"github.com/krystal/go-katapult"
+)
+
+type SecurityGroupRule struct {
+	ID        string    `json:"id,omitempty"`
+	Direction string    `json:"direction,omitempty"`
+	Protocol  string    `json:"protocol,omitempty"`
+	Ports     *string   `json:"ports,omitempty"`
+	Targets   *[]string `json:"targets,omitempty"`
+	Notes     *string   `json:"notes,omitempty"`
```
Comment (jimeh) [inline core/security_group_rule.go:17]: For this struct these fields actually don't need to be pointer types, as it's used to hold rule information received from the API. It's only the `SecurityGroupRuleArguments` struct where they need to be pointers so we can differentiate in JSON between not sending the key at all or explicitly setting it to empty string/array when creating/updating rules.

**Review by jimeh — CHANGES_REQUESTED** [review]

**Comment by andrewjtait** [comment]
@jimeh also ready for re-review 🙏

**Review by jimeh — APPROVED** [review]

**Pull request merged**